### DISCUSSION TITLE
How to extend query options

### DISCUSSION THREAD

**Discussion by SwapnilSoni1999** [prisma/prisma#18966] [Q&A]
Prisma supports middleware functions, Here we can access `args` prop but I need some support to pass custom options

for eg.
```js
prisma.users.update({
    where: { id: myId },
    data: { some: 'value' },
    custom: 'option' // this one
})
```

so that I can use it in my middleware

```js
prisma.$use((params, next) => {
     // I wanna use 'custom'
     params.args.custom
})
```

here the problem is typescript types. If I pass `custom` in query it will throw an error

**Comment by ludralph** [comment]
Hi @SwapnilSoni1999 👋 

Have you taken a look at Prisma [Client extensions](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn)? Client extensions should be able to solve your usecase. The Query component lets you hook into the query lifecycle and perform side effects, modify query arguments, or alter the results in a type-safe way. These are an alternative to [middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) that provide full type safety.

**Reply by SwapnilSoni1999** [reply]
> Hi @SwapnilSoni1999 wave
> 
> Have you taken a look at Prisma [Client extensions](https://www.prisma.io/blog/client-extensions-preview-8t3w27xkrxxn)? Client extensions should be able to solve your usecase. The Query component lets you hook into the query lifecycle and perform side effects, modify query arguments, or alter the results in a type-safe way. These are an alternative to [middleware](https://www.prisma.io/docs/concepts/components/prisma-client/middleware) that provide full type safety.

Hey @ludralph thank you for responding. 
I actually want to access the custom property in middleware and based on that I want to modify query args. The problem is I tried passing it in default `.findMany` query and I'm getting an error 

```
Invalid `prisma.roles.findMany()` invocation:

{
  where: {
    deletedAt: {
      not: null
    },
    id: '1'
  },
  bypassDeleted: true
  ~~~~~~~~~~~~~
}

Unknown arg `bypassDeleted` in bypassDeleted for type Roles. Did you mean `select`? Available args:
type findManyRoles {
  where?: RolesWhereInput
  orderBy?: List<RolesOrderByWithRelationInput> | RolesOrderByWithRelationInput
  cursor?: RolesWhereUniqueInput
  take?: Int
  skip?: Int
  distinct?: List<RolesScalarFieldEnum>
}

```

Can you please give me some example how do I pass `bypassDeleted` prop so I can access that in middleware `$use` function

**Reply by MaheeGamage** [reply]
I also looking for a way to add a custom input property to the query. I have created a custom multitenant extension and sometimes I need to disable tenantId injection to the query

```
const updateUser = await prisma.user.update({
  where: {
    email: 'viola@prisma.io',
  },
  data: {
    name: 'Viola the Magnificent',
  },
  skipTenantInjection: false;
})
```

And access this value in multitenant extension

```
Prisma.defineExtension((client) => {
    return client.$extends({
      query: {
        $allModels: {
          async $allOperations({ model, operation, args, query }) {
            const skipTenantInjection = args.skipTenantInjection;
          }
        }
      }
    })
})
```

Let me know if someone finds a way to achieve this

**Reply by bambrose24** [reply]
@MaheeGamage I've been looking for this as well, based on seeing the `@prisma/extension-accelerate` do this successfully, but I can't seem to find where the package's source is actually defined.

**Reply by outofthisworld** [reply]
Defined in a d.ts file like so...! `( looks super straight forward ;) )`

```
export declare function makeAccelerateExtension(fetcher: PrismaFetch): (client: any) => PrismaDefault.PrismaClientExtends<Types.Extensions.InternalArgs<{}, {
    $allModels: {
        aggregate<T, A>(this: T, args: Prisma.Exact<A, Prisma.Args<T, "aggregate"> & PrismaCacheStrategy>): AcceleratePromise<Prisma.Result<T, A, "aggregate">>;
        count<T_1, A_1>(this: T_1, args?: Prisma.Exact<A_1, Prisma.Args<T_1, "count"> & PrismaCacheStrategy> | undefined): AcceleratePromise<Prisma.Result<T_1, A_1, "count">>;
        findFirst<T_2, A_2>(this: T_2, args?: Prisma.Exact<A_2, Prisma.Args<T_2, "findFirst"> & PrismaCacheStrategy> | undefined): AcceleratePromise<Prisma.Result<T_2, A_2, "findFirst"> | null>;
        findFirstOrThrow<T_3, A_3>(this: T_3, args?: Prisma.Exact<A_3, Prisma.Args<T_3, "findFirstOrThrow"> & PrismaCacheStrategy> | undefined): AcceleratePromise<Prisma.Result<T_3, A_3, "findFirstOrThrow">>;
        findMany<T_4, A_4>(this: T_4, args?: Prisma.Exact<A_4, Prisma.Args<T_4, "findMany"> & PrismaCacheStrategy> | undefined): AcceleratePromise<Prisma.Result<T_4, A_4, "findMany">>;
        findUnique<T_5, A_5>(this: T_5, args: Prisma.Exact<A_5, Prisma.Args<T_5, "findUnique"> & PrismaCacheStrategy>): AcceleratePromise<Prisma.Result<T_5, A_5, "findUnique"> | null>;
        findUniqueOrThrow<T_6, A_6>(this: T_6, args: Prisma.Exact<A_6, Prisma.Args<T_6, "findUniqueOrThrow"> & PrismaCacheStrategy>): AcceleratePromise<Prisma.Result<T_6, A_6, "findUniqueOrThrow">>;
        groupBy<T_7, A_7>(this: T_7, args: Prisma.Exact<A_7, Prisma.Args<T_7, "groupBy"> & PrismaCacheStrategy>): AcceleratePromise<Prisma.Result<T_7, A_7, "groupBy">>;
    };
}, {}, {}> & Types.Extensions.InternalArgs<{}, {}, {}, {}> & Types.Extensions.DefaultArgs>;
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Reply by IanvanZyl636** [reply]
Hi @ludralph , please open discussion again as this is unanswered

**Comment by IanvanZyl636** [comment]
I am also stuck with the same issue @SwapnilSoni1999  @MaheeGamage @bambrose24  have you guys find a clean implementation for this issue to make all models and all operators type safe with newly added property?
